src/csharp/Microsoft.Spark/Sql/Types/DataType.cs

```diff
@@ -39,10 +40,13 @@ public abstract class DataType
             typeof(MapType),
             typeof(StructType) };
 
+        private static string[] s_simpleTypeNormalizedNames = null;
+        private static string[] s_complexTypeNormalizedNames = null;
+
         /// <summary>
         /// Normalized type name.
         /// </summary>
-        public string TypeName => NormalizeTypeName(GetType().Name);
+        public string TypeName => NormalizeTypeName(GetType());
```

Comment:
Instead of having a new method for this, it may be simpler to normalize the name and cache the result. Something like

```csharp
        private string _typeName;

        /// <summary>
        /// Normalized type name.
        /// </summary>
        public string TypeName
        {
            get
            {
                return _typeName ?? (_typeName = NormalizeTypeName(GetType().Name));
            }
        }
```